safe/gui/tools/minimum_needs/needs_calculator_dialog.py

```diff
@@ -61,148 +96,107 @@ def __init__(self, parent=None):
-        layer_id = self.cboPolygonLayers.itemData(
-            index, QtCore.Qt.UserRole)
-        # noinspection PyArgumentList
-        layer = QgsMapLayerRegistry.instance().mapLayer(layer_id)
-        fields = layer.pendingFields()
-        self.cboFields.clear()
-        has_fields = False
-        for field in fields:
-            # LOGGER.info(field.typeName())
-            # TODO exclude dates too? TS
-            if field.typeName() != 'String':
-                has_fields = True
-                add_ordered_combo_item(
-                    self.cboFields, field.name(), field.name())
-
-        # Now disable the run button if no suitable fields were found
-        # see #2206
-        ok_button = self.button_box.button(QtGui.QDialogButtonBox.Ok)
-        if not has_fields:
-            ok_button.setEnabled(False)
-        else:
-            ok_button.setEnabled(True)
+        # create memory layer
+        output_layer = (
+            '%s_minimum_needs' % input_layer.name())
+        output_layer = create_memory_layer(
+            output_layer,
+            input_layer.geometryType(),
+            input_layer.crs(),
+            input_layer.fields())
+
+        # monkey patching input layer to make it work with
+        # prepare vector layer function
+        temp_layer = input_layer
+        temp_layer.keywords = {
+            'layer_purpose': layer_purpose_aggregation['key']}
+
+        # add keywords to output layer
+        copy_layer(temp_layer, output_layer)
+
+        return output_layer
 
     def accept(self):
         """Process the layer and field and generate a new layer.
 
         .. note:: This is called on OK click.
 
         """
-        index = self.cboFields.currentIndex()
-        field_name = self.cboFields.itemData(
-            index, QtCore.Qt.UserRole)
+        # Create a new layer for output layer
+        output_layer = self.prepare_new_layer(self.layer.currentLayer())
 
-        index = self.cboPolygonLayers.currentIndex()
-        layer_id = self.cboPolygonLayers.itemData(
-            index, QtCore.Qt.UserRole)
-        # noinspection PyArgumentList
-        layer = QgsMapLayerRegistry.instance().mapLayer(layer_id)
+        # Monkey patching output layer to make it work with
+        # minimum needs calculator
+        output_layer.keywords['layer_purpose'] = (
+            layer_purpose_aggregation['key'])
+        output_layer.keywords['inasafe_fields'] = (
+            {displaced_field['key']: self.displaced.currentField(),
+             aggregation_name_field['key']:
+                 self.aggregation_name.currentField()})
 
-        file_name = layer.source()
+        # remove unnecessary fields & rename inasafe fields
+        rename_remove_inasafe_fields(output_layer)
 
         try:
-            output_layer = self.minimum_needs(input_layer, field_name)
-        except ValueError:
+            self.minimum_needs(output_layer)
+        except Exception as e:
+            LOGGER.debug(e)
```

Comment:
Logging the error is not enough. We should also display a message box to the user.
The best would be to display an error based on the exception.
You can at least use `safe.utilities.qgis_utilities.display_critical_message_box' to say that something went wrong.